include/flux/op/cartesian_base.hpp

```diff
@@ -0,0 +1,359 @@
+
+template<typename... Ts>
+struct require_single_type {
+};
+
+template<typename T>
+struct require_single_type<T> {
+    using type = T;
+};
+
+template<typename... Ts>
+using require_single_type_t = typename require_single_type<Ts...>::type;
+
+template <typename T, std::size_t RepeatCount>
+struct tuple_repeated {
+    constexpr static auto repeat_tuple(T value) {
+        return [value]<std::size_t... Is>(std::index_sequence<Is...>) {
+            return std::tuple{(static_cast<void>(Is), value)...};
+        }(std::make_index_sequence<RepeatCount>{});
+    }
+
+    using type = decltype(repeat_tuple(std::declval<T>()));
+};
+
+template <typename T, std::size_t RepeatCount>
+using tuple_repeated_t = tuple_repeated<T, RepeatCount>::type;
+
+template<std::size_t Arity, cartesian_kind CartesianKind, read_kind ReadKind, typename... Bases>
+struct cartesian_traits_types {
+};
+
+template<std::size_t Arity, typename Base>
+struct cartesian_traits_types<Arity, cartesian_kind::power, read_kind::tuple, Base> {
+    using value_type = tuple_repeated_t<value_t<Base>, Arity>;
+};
+
+template<std::size_t Arity, typename... Bases>
+struct cartesian_traits_types<Arity, cartesian_kind::product, read_kind::tuple, Bases...> {
+    using value_type = std::tuple<value_t<Bases>...>;
+};
+
+template <std::size_t Arity, cartesian_kind CartesianKind, read_kind ReadKind, typename... Bases>
+struct cartesian_traits_base_impl {
+private:
+
+    template<std::size_t I, typename Self>
+    requires (CartesianKind == cartesian_kind::power)
+    static constexpr auto&& get_base(Self& self) {
+        return self.base_;
+    }
+
+    template<std::size_t I, typename Self>
+    requires (CartesianKind == cartesian_kind::product)
+    static constexpr auto&& get_base(Self& self) {
+        return std::get<I>(self.bases_);
+    }
+
+
+    template <std::size_t I, typename Self>
+    static constexpr auto inc_impl(Self& self, cursor_t<Self>& cur) -> cursor_t<Self>&
+    {
+        flux::inc(get_base<I>(self), std::get<I>(cur));
+
+        if constexpr (I > 0) {
+            if (flux::is_last(get_base<I>(self), std::get<I>(cur))) {
+                std::get<I>(cur) = flux::first(get_base<I>(self));
+                inc_impl<I-1>(self, cur);
+            }
+        }
+
+        return cur;
+    }
+
+
+    template <std::size_t I, typename Self>
+    static constexpr auto dec_impl(Self& self, cursor_t<Self>& cur) -> cursor_t<Self>&
+    {
+        if (std::get<I>(cur) == flux::first(get_base<I>(self))) {
+            std::get<I>(cur) = flux::last(get_base<I>(self));
+            if constexpr (I > 0) {
+                dec_impl<I-1>(self, cur);
+            }
+        }
+
+        flux::dec(get_base<I>(self), std::get<I>(cur));
+
+        return cur;
+    }
+
+    template <std::size_t I, typename Self>
+    static constexpr auto ra_inc_impl(Self& self, cursor_t<Self>& cur, distance_t offset)
+    -> cursor_t<Self>&
+    {
+        if (offset == 0)
+            return cur;
```

Comment:
Style: braces after if statements please 
```suggestion
        if (offset == 0) {
            return cur;
        }
```